bootstrap.md

```diff
@@ -0,0 +1,136 @@
+
+This is obviously impractical, so we choose to start the bootstrap sequence with nodes by connecting with an inital set of trusted peers. These can be nodes run by well-known entities (the Filecoin foundation, large exchanges, good actors in the community, etc) or manually specified by the node operator in the config file. We leave trustless (or fully decentralized) bootstrapping for future work.
+
+A strawman bootstrapping sequence would have nodes use these initial peers to sync up to the canonical chain. However, new connection requests might effecitvely DoS these peers, making the Filecoin network unavailable to new nodes. Therefore, Filecoin bootstrapping seeks to "load-balance" initial requests to ensure the network remains widely available.
+
+We connect to each of these initial peers using the `hello handshake` (see in [network protocols](./network-protocols.md#hello-handshake)), getting back
+
+```go
+type HelloMessage struct {
+HeaviestTipSet []Cid
+HeaviestTipSetHeight uint64
+GenesisHash Cid
+}
+```
+
+followed by a `GetPeers` request, getting back
+
+```go
+type PeersMessage struct {
+PeerSet []PeerID
+}
+```
+
+We will use these new peers to sync with the chain, using the chain heads from our initial set of trusted nodes as a means of verifying that these peers are mining the canonical chain. These initial peers do not return their latest heads, but rather their latest confirmed heads, thereby ensuring other honest peers should be aware of these as well.
+
+We assume no long-lived network partitions in Filecoin, meaning these confirmed heads will all be on the same chain (though potentially at different heights). A node should initially connect to 5 trusted nodes to begin bootstrapping.
+
+```go
+type Bootstrapper struct {
+    trustedPeers []PeerID
+    trustedHeads map[[]Cid]uint64
+    genesisHash Cid
+    
+    bootstrapPeers []PeerID
+    
+}
+
+func InitialConnect(config ConfigFile, b Bootstrapper, s Syncer) {
+    b.trustedPeers = configFile.bootstrapPeers
+    
+    for trustedPeer := range b.trustedPeers {
+        trustedChain := sayHello(trustedPeer)
+        otherPeers := getPeers(trustedPeer).PeerSet
+        
+        b.trustedHeads[trustedChain.HeaviestTipSet] = trustedChain.HeaviestTipSetHeight
+    	
+        if b.genesisHash == nil {
+	        b.genesisHash = trustedChain.GenesisHash           
+        } else if b.genesisHash != trustedChain.GenesisHash {
+            panic("One of your trusted peers is bad.")
+        }
+        
+        b.bootstrapPeers = append(b.bootstrapPeers, otherPeers)
+    }
+    
+    s.genesis = b.genesisHash
```

Comment:
note: we will know our expected genesis block before starting bootstrapping

Reply:
I'm not a blockchain person, but that sounds like something you will publish and will be universally known once it goes into production right? Kind of like the empty hash, or a seed.

Reply:
Yes, that is right @anacrolix. My thought initially @whyrusleeping was that your set of trusted peers may actually determine which chain you want to sync to, as in if they collectively decide to abandon a Filecoin for another, you'd want to follow them. But that may clearly be the wrong trust model (you see, I'm a Stellar fanboy).

Should I just hardcode GENESIS in there?

Reply:
Yeah, hardcode it.

Also, I'm pretty sure that even in stellar what you describe cannot happen.